proposals/nnnn-rpc-design-refactoring.md

```diff
@@ -0,0 +1,149 @@
+ 
+#### Provide plugin architecture for adding additional functionality and RPC groups easily.
+
+SDL RPC's can be split into groups of RPC's related to certain functionality (for example RC or video streaming).  
+These RPC's may work with functionality that is not required by other RPC's, so exposure should be limited to the RPC's group.
+
+Adding new RPC or groups of RPC's is currently a very complicated and unclear process.
+
+The best solution is to adopt a plugin approach of adding new RPC's to SDL core.
+This will allow OEM's to implement code for new RPC's with ease by implementing certain interfaces and providing them to SDL as shared library.
+
+
+## Detailed design
+
+### Use dependency injection for commands 
+
+Commands should be accepted as a list of services that they require. The mobile factory should be responsible for providing this list of dependencies.  
+
+#### Current Approach 
+```cpp
+
+SomeCommand::SomeCommand(app_mngr);
+
+void SomeCommand::Run() {
+  application_manager->protocol_handler().do_stuff();
+  application_manager->state_controller().do_stuff();
+  application_manager->PolicyHandler().do_stuff();
+  application_manager->resume_controller().do_stuff();
+  
+}
+```
+
+#### New Approach 
+```cpp
+
+SomeCommand::SomeCommand(protocol_handler, state_controller, policy_handler, resume_controller);
+
+void SomeCommand::Run() {
+  protocol_handler_.do_stuff();
+  state_controller_.do_stuff();
+  protocol_handler_.do_stuff();
+  protocol_handler_.do_stuff();
+}
+```
+This will simplify code and provide the possibility to add a new component to certain commands without affecting other commands.
+The factory will create the command with all required parameters, so existing code will not be complicated.
+RC component contains class ResourceAllocationManager. And non-RC commands should have no access to this class, so it should be passed only to RC commands. 
+
+### Reduce responsibility of ApplicationManager class 
+
+Logic of handling and sending RPC's should be extracted from ApplicationManager to separate components `RPCHandler`, `RPCService`.
+
+### RPCHandler
+RPCHandler class should implement interfaces ```HMIMessageObserver```, ```ProtocolObserver```, handle incoming RPC's from HMI or mobile 
+and process them. 
+
+RPCHandler responsibilities:
+ - Handle HMI/Mobile RPC's in message loop
+ - Class checking policies on each RPC
+ - Creating command from message with Command Factories
+ - Send command for execution to Request Controller
+ 
+ ### RPCService
+
+RPCService should provide interface for sending messages to Mobile/HMI.  
+RPCService should be passed to commands for Sending RPC's using dependency injection approach.
+
+RPCService responsibilities:
+ - SendMessage to HMI/Mobile
+ - Call checking policies if message is allowed to send
+ - Provide Request Controller information of any incoming request should be terminated.  
```

Comment:
Please update to either "...information if any incoming request..." or "...information of any incoming request that should..." depending on the intended meaning of this responsibility.